File: domain/src/main/java/home/oleg/placenearme/interactors/CheckInOut.java
package home.oleg.placenearme.interactors;

import home.oleg.placenearme.models.DetailedVenue;
import home.oleg.placenearme.repositories.VenueHistoryRepository;
import io.reactivex.Completable;

public class CheckInOut {

    private final VenueHistoryRepository venueHistoryRepository;

    public CheckInOut(VenueHistoryRepository venueHistoryRepository) {
        this.venueHistoryRepository = venueHistoryRepository;
    }

    public Completable execute(DetailedVenue detailedVenue) {
        if (detailedVenue.isHereNow()) {
            return venueHistoryRepository.checkOut(detailedVenue.getId());
        } else {
            return venueHistoryRepository
                    .checkOutFromCurrent()
                    .andThen(venueHistoryRepository.checkIn(detailedVenue));
        }
    }

}

File: feature-add-history/src/main/java/home/oleg/placesnearme/feature_add_history/domain/interactor/CheckInOut.kt
package home.oleg.placesnearme.feature_add_history.domain.interactor

import home.oleg.placesnearme.coredomain.models.Place
import home.oleg.placesnearme.coredomain.repositories.VenueHistoryRepository
import io.reactivex.Single
import javax.inject.Inject

class CheckInOut @Inject constructor(private val venueHistoryRepository: VenueHistoryRepository) {

    fun execute(detailedVenue: Place): Single<Boolean> {
        return if (detailedVenue.isHereNow) {
            venueHistoryRepository.dropCheckIns()
                    .andThen(Single.just(false))
        } else {
            venueHistoryRepository.checkIn(detailedVenue)
                    .andThen(Single.just(true))
        }
    }

}
